presto-main/src/main/java/io/prestosql/sql/planner/DesugarRowSubscriptRewriter.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.Session;
+import io.prestosql.spi.type.RowType;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.tree.Cast;
+import io.prestosql.sql.tree.DereferenceExpression;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.ExpressionRewriter;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.Identifier;
+import io.prestosql.sql.tree.LongLiteral;
+import io.prestosql.sql.tree.NodeRef;
+import io.prestosql.sql.tree.SubscriptExpression;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Replaces subscript expression on Row
+ * with cast and dereference:
+ * <pre>
+ *     ROW (1, 'a', 2) [2]
+ * </pre>
+ * is transformed into:
+ * <pre>
+ *     (CAST (ROW (1, 'a', 2) AS ROW (field_0 bigint, field_1 varchar(1), field_2 bigint))).field_1
+ * </pre>
+ */
+public class DesugarRowSubscriptRewriter
+{
+    private DesugarRowSubscriptRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, TypeAnalyzer typeAnalyzer, SymbolAllocator symbolAllocator)
+    {
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        /*if (expression instanceof SymbolReference) {
+            return expression;
+        }*/
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression);
+
+        return rewrite(expression, expressionTypes);
+    }
+
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes), expression);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+        }
+
+        @Override
+        public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression base = treeRewriter.rewrite(node.getBase(), context);
+            Expression index = treeRewriter.rewrite(node.getIndex(), context);
+
+            Type type = getType(base);
```

Comment:
If `node.getBase()` is rewritten by the call above, it won't have a type registered in `expressionTypes` and this will return null.

You can see that with a query like:

```sql
SELECT x[1][1] FROM (VALUES row(row(row(1, 2)))) t(x);
```

It works only because null is not an instance of any type and the rewriter gets applied multiple times, so it first transforms:

```
x[1][1]
```
=>
```
CAST("field_1" AS row(field_0 row(integer,integer))).field_0[1]
```

and then

```
CAST(CAST("field" AS row(field_0 row(integer,integer))).field_0 AS row(field_0 integer,field_1 integer)).field_0
```

Reply:
Yes, it was naive to check `getType(base)` and it killed the recursion.